client/grpc/simulate/simulate.go

```diff
@@ -0,0 +1,101 @@
+package simulate
+
+import (
+	"context"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	"github.com/cosmos/cosmos-sdk/client"
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	txtypes "github.com/cosmos/cosmos-sdk/types/tx"
+	"github.com/cosmos/cosmos-sdk/types/tx/signing"
+	authtx "github.com/cosmos/cosmos-sdk/x/auth/tx"
+)
+
+type simulateServer struct {
+	app         baseapp.BaseApp
+	pubkeyCodec cryptotypes.PublicKeyCodec
+	txConfig    client.TxConfig
+}
+
+// NewSimulateServer creates a new SimulateServer.
+func NewSimulateServer(app baseapp.BaseApp, pubkeyCodec cryptotypes.PublicKeyCodec, txConfig client.TxConfig) SimulateServiceServer {
+	return simulateServer{
+		app:         app,
+		pubkeyCodec: pubkeyCodec,
+		txConfig:    txConfig,
+	}
+}
+
+var _ SimulateServiceServer = simulateServer{}
+
+// Simulate implements the SimulateService.Simulate RPC method.
+func (s simulateServer) Simulate(ctx context.Context, req *SimulateRequest) (*SimulateResponse, error) {
+	if req.Tx == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid empty tx")
+	}
+
+	req.Tx.UnpackInterfaces(s.app.GRPCQueryRouter().AnyUnpacker())
+	txBuilder, err := txBuilderFromProto(s.txConfig, s.pubkeyCodec, req.Tx)
+	if err != nil {
+		return nil, err
+	}
+
+	txBytes, err := req.Tx.Marshal()
+	if err != nil {
+		return nil, err
+	}
+
+	gasInfo, result, err := s.app.Simulate(txBytes, txBuilder.GetTx())
+	if err != nil {
+		return nil, err
+	}
+
+	return &SimulateResponse{
+		GasInfo: &gasInfo,
+		Result:  result,
+	}, nil
+}
+
+// txBuilderFromProto converts a proto.Message Tx into a TxBuilder.
+func txBuilderFromProto(txConfig client.TxConfig, pubkeyCodec cryptotypes.PublicKeyCodec, tx *txtypes.Tx) (client.TxBuilder, error) {
```

Comment:
If someone has a better idea on how to convert a tx.Tx into a TxBuilder, I'm happy to hear it.

Reply:
Sorry if this is obvious, but why do we need to do this? It seems like a codesmell -- a builder is used to make/sign a tx, not the other way around.

Reply:
yeah, same feeling as you, that's why I wrote this comment.

The Simulate endpoint takes a `tx.Tx` (proto message). Then `baseapp.Simulate()` takes a `sdk.Tx`. So we need to convert from the first one to the second one:
- the path I chose here where to convert `tx.Tx` to a txBuilder, then use `txBuilder.GetTx()`
- the second path is to let `tx.Tx` (the proto message) implement `sdk.Tx`. This would require adding methods such as ValidateBasic, GetMsgs on `tx.Tx`, probably stuff like GetGas too.

I'll try to have a look at the 2nd path.

Reply:
Seems like the 2nd approach is better, but I'm not clear on the LOE of that -- hopefully it's not too difficult. Thoughts @aaronc? Another thing to consider is the refactoring of `baseapp.Simulate()` to take a `tx.Tx` instead -- aren't we deprecating `sdk.Tx` anyway?

Reply:
I explained the approach to @amaurymartiny this morning. Basically we need a simple constructor to wrap in `builder`. Incoming tx's should always be wrapped because direct sign bytes needs to preserve raw encoded bytes. It's not relevant for simulation but that's why `tx.Tx` doesn't implement `sdk.Tx`

Reply:
> we need a simple constructor to wrap in `builder`

I'm sorry, but I don't understand what this means.

Reply:
Have a look at `WrapTxBuilder`. @aaronc lmk if that's what you had in mind.

Reply:
Yes, we can maybe even just call it `WrapTx`. Basically because we need to retain the raw body and auth_info bytes from `TxRaw`, we need a wrapper type around `tx.Tx` which retains those encoding details. Currently that type is just called `builder`, but it should maybe be called `wrapper`.